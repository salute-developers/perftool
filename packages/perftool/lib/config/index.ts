export { getConfig } from './common';
export type { Config, ProjectConfig } from './common';

export { getTaskConfig, getAllTasks } from './task';

export { getWebpackConfig } from './webpack';

export { buildDirectory, sourceDirectory } from './paths';
