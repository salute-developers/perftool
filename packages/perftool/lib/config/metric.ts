import nativeMetrics from '../statistics/metrics';

import type { Config } from './common';

export const defaultMetricConfiguration = {
    median: {
        failOnSignificantChanges: false,
        enable: false,
    },
    gMean: {
        failOnSignificantChanges: false,
        enable: false,
    },
    hMean: {
        failOnSignificantChanges: false,
        enable: false,
    },
    iqr: {
        failOnSignificantChanges: false,
        enable: false,
    },
    idr: {
        failOnSignificantChanges: false,
        enable: false,
    },
    midhinge: {
        failOnSignificantChanges: false,
    },
    trimean: {
        failOnSignificantChanges: false,
    },
    truncMean5: {
        failOnSignificantChanges: false,
        enable: false,
    },
    truncMean10: {
        failOnSignificantChanges: false,
        enable: false,
    },
    truncMean25: {
        failOnSignificantChanges: false,
        enable: false,
    },
    hl: {
        failOnSignificantChanges: false,
    },
};

export function getAllMetrics(config: Pick<Config, 'metricConfiguration' | 'metrics'>, filterDisabled = true) {
    return [...config.metrics, ...nativeMetrics].filter((metric) => {
        if (!filterDisabled) return true;

        const metricConfig = config.metricConfiguration[metric.id] || {};

        return metricConfig.enable === undefined ? true : metricConfig.enable;
    });
}
