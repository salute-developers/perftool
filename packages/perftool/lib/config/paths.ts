import path from 'path';

import CWD from '../utils/cwd';

export const mainDirectory = path.resolve(CWD, './.perftool');

export const buildDirectory = path.resolve(mainDirectory, 'build');

export const sourceDirectory = path.resolve(mainDirectory, 'lib');

export const cacheDirectory = path.resolve(mainDirectory, 'cache');
