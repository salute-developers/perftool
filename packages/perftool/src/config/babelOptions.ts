import * as babelCore from '@babel/core';

export const babelOptions: babelCore.TransformOptions = {
    sourceType: 'module',
    presets: [
        ['@babel/preset-env', { targets: { chrome: '90', esmodules: true } }],
        '@babel/preset-react',
        '@babel/preset-typescript',
    ],
    plugins: [],
};
