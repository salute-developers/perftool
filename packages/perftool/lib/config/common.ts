import os from 'os';
import { Configuration as WebpackConfig } from 'webpack';
import { PuppeteerNodeLaunchOptions } from 'puppeteer';
import merge from 'deepmerge';

import { Task, MeasurerConfig } from '../client/measurement/types';
import { ExportPickRule } from '../build/collect';
import { Metric } from '../statistics/types';
import { debug } from '../utils/logger';

import { cacheDirectory } from './paths';

type MetricConfiguration = {
    enable?: boolean;
    confidenceLevel?: number;
    failOnSignificantChanges?: boolean;
};

type TaskConfiguration = MeasurerConfig & {
    enable?: boolean;
    include?: string[];
    exclude?: string[];
    retries?: number;
    failOnSignificantChanges?: boolean;
    metrics?: {
        include?: string[];
        exclude?: string[];
    };
};

// TODO add comments
export type Config = {
    taskConfiguration: {
        [key: string]: TaskConfiguration;
    };
    tasks: Task<any, any>[];
    metricConfiguration: {
        [key: string]: MetricConfiguration;
    };
    stabilizers: string[];
    absoluteError: number;
    metrics: Metric<any>[];
    include: string[];
    exclude: string[];
    jobs: number;
    retries: number;
    baseBranchRef?: string;
    currentBranchRef?: string;
    cache: {
        taskState?: boolean;
    };
    cacheDirectory: string;
    cacheExpirationTime: number;
    displayIntermediateCalculations: boolean;
    intermediateRefreshInterval: number;
    failOnSignificantChanges: boolean;
    outputFilePath: string;
    configPath?: string;
    logLevel: 'quiet' | 'normal' | 'verbose';
    puppeteerOptions: PuppeteerNodeLaunchOptions;
    taskWaitTimeout: number;
    runWaitTimeout: number;
    dryRunTimes: number;
    modifyWebpackConfig: (defaultConfig: WebpackConfig) => WebpackConfig;
    exportPickRule: ExportPickRule;
};

export type ProjectConfig = Partial<Omit<Config, 'configPath' | 'logLevel'>>;

export type CliConfig = Partial<
    Pick<
        Config,
        | 'include'
        | 'configPath'
        | 'outputFilePath'
        | 'logLevel'
        | 'failOnSignificantChanges'
        | 'baseBranchRef'
        | 'currentBranchRef'
    >
>;

function withDefault<T>(value: T | undefined, defaultValue: T): T {
    if (typeof value === 'undefined') {
        return defaultValue;
    }

    return value;
}

export function getConfig(cliConfig: CliConfig = {}, projectConfig: ProjectConfig = {}): Config {
    debug('getting final config');
    const mixedInputConfig = merge(projectConfig, cliConfig);

    const result = {
        taskConfiguration: withDefault(mixedInputConfig.taskConfiguration, {}),
        tasks: withDefault(mixedInputConfig.tasks, []),
        metricConfiguration: withDefault(mixedInputConfig.metricConfiguration, {}),
        stabilizers: withDefault(mixedInputConfig.stabilizers, []),
        absoluteError: withDefault(mixedInputConfig.absoluteError, 0),
        metrics: withDefault(mixedInputConfig.metrics, []),
        include: withDefault(mixedInputConfig.include, []),
        exclude: withDefault(mixedInputConfig.exclude, []),
        jobs: withDefault(mixedInputConfig.jobs, Math.max(os.cpus().length - 1, 1)),
        retries: withDefault(mixedInputConfig.retries, 10),
        baseBranchRef: withDefault(mixedInputConfig.baseBranchRef, undefined),
        currentBranchRef: withDefault(mixedInputConfig.currentBranchRef, undefined),
        cache: withDefault(mixedInputConfig.cache, {}),
        cacheDirectory: withDefault(mixedInputConfig.cacheDirectory, cacheDirectory),
        cacheExpirationTime: withDefault(mixedInputConfig.cacheExpirationTime, 0),
        displayIntermediateCalculations: withDefault(mixedInputConfig.displayIntermediateCalculations, true),
        intermediateRefreshInterval: withDefault(mixedInputConfig.intermediateRefreshInterval, 10000),
        failOnSignificantChanges: withDefault(mixedInputConfig.failOnSignificantChanges, true),
        outputFilePath: withDefault(mixedInputConfig.outputFilePath, 'perftest/report-[time].json'),
        configPath: withDefault(mixedInputConfig.configPath, undefined),
        logLevel: withDefault(mixedInputConfig.logLevel, 'normal'),
        puppeteerOptions: withDefault(mixedInputConfig.puppeteerOptions, {}),
        taskWaitTimeout: withDefault(mixedInputConfig.taskWaitTimeout, 1000 * 10),
        runWaitTimeout: withDefault(mixedInputConfig.runWaitTimeout, 1000 * 60 * 2),
        dryRunTimes: withDefault(mixedInputConfig.taskWaitTimeout, 1),
        modifyWebpackConfig: withDefault(mixedInputConfig.modifyWebpackConfig, (c) => c),
        exportPickRule: withDefault(mixedInputConfig.exportPickRule, 'named'),
    };

    debug('final config: ', result);

    return result;
}
