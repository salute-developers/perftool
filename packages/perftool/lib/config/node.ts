import path from 'path';

import checkPath from '../utils/checkPath';
import { debug, error, info } from '../utils/logger';
import CWD from '../utils/cwd';

import { ProjectConfig } from './common';

const defaultJsConfigPath = path.resolve(CWD, './perftool.config.mjs');
const defaultTsConfigPath = path.resolve(CWD, './perftool.config.mts');

type ImportedConfig = {
    path: string;
    value: ProjectConfig;
};

export async function importConfig(cliConfigPath?: string): Promise<ImportedConfig | null> {
    debug('importing project config');
    let configPath = cliConfigPath ? path.resolve(CWD, cliConfigPath) : undefined;
    debug('cli config path is ', configPath);

    if (configPath && !(await checkPath(configPath))) {
        error('provided config path is inaccessible');
        throw new Error();
    }

    if (!configPath && (await checkPath(defaultJsConfigPath))) {
        debug('no provided config path, using default ', defaultJsConfigPath);
        configPath = defaultJsConfigPath;
    }

    if (!configPath && (await checkPath(defaultTsConfigPath))) {
        debug('no provided config path, using default ', defaultTsConfigPath);
        configPath = defaultTsConfigPath;
    }

    if (!configPath) {
        debug('no project config found, using default config');
        return null;
    }

    info('Using project config', configPath);

    const { default: value } = await import(configPath);

    debug('import succeeded, project config:', value);

    return {
        value,
        path: configPath,
    };
}
