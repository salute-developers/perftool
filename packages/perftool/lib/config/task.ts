import merge from 'deepmerge';

import { Task } from '../client/measurement/types';
import nativeTasks from '../client/measurement';

import { Config } from './common';

type TaskConfig<T extends Task<any, any>> = Parameters<T['run']>[0]['config'];

export function getTaskConfig<T extends Task<any, any>>(task: T, config: Config): TaskConfig<T> {
    return merge(task.defaultConfig, config.taskConfiguration[task.id] || {});
}

export function getAllTasks(config: Config) {
    return [...config.tasks, ...nativeTasks].filter((task) => {
        const taskConfig = getTaskConfig(task, config);

        return taskConfig.enable === undefined ? true : taskConfig.enable;
    });
}
