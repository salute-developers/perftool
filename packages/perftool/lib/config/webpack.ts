import HtmlWebpackPlugin from 'html-webpack-plugin';
import webpack, { Configuration as WebpackConfig } from 'webpack';
import { createRequire } from 'node:module';
import path from 'path';
import { fileURLToPath } from 'url';

import { debug } from '../utils/logger';
import CWD from '../utils/cwd';

import { Config } from './common';

const require = createRequire(import.meta.url);
const dirname = path.dirname(fileURLToPath(import.meta.url));
const ownNodeModules = path.relative(CWD, path.resolve(dirname, '../../node_modules'));

const defaultConfig: WebpackConfig = {
    mode: 'production',
    context: CWD,
    resolve: {
        extensions: ['.ts', '.tsx', '.js', '.jsx', '.mjs'],
        modules: ['node_modules', ownNodeModules],
        fallback: {
            'react-dom/client': false,
        },
    },
    experiments: {
        topLevelAwait: true,
    },
    performance: {
        hints: false,
        maxAssetSize: 1024 * 1024,
        maxEntrypointSize: 1024 * 1024,
    },
    output: {
        filename: 'bundle.[chunkhash].js',
    },
    module: {
        rules: [
            {
                loader: require.resolve('babel-loader'),
                test: /\.(js|mjs|jsx|ts|tsx)$/,
                exclude: /node_modules/,
                options: {
                    presets: [
                        ['@babel/preset-env', { targets: { chrome: '90', esmodules: true } }],
                        '@babel/preset-react',
                        '@babel/preset-typescript',
                    ],
                    plugins: [],
                },
            },
        ],
    },
    plugins: [
        new HtmlWebpackPlugin(),
        new webpack.DefinePlugin({
            'process.env.PERFTOOL_CLIENT_RUNTIME': JSON.stringify(true),
        }),
    ],
};

export function getWebpackConfig(entry: string, output: string, config: Config): WebpackConfig {
    debug('creating webpack config');

    const finalConfig = config.modifyWebpackConfig(defaultConfig);

    finalConfig.entry = entry;
    finalConfig.output!.path = output;

    debug('final webpack config', JSON.stringify(finalConfig, null, 2));

    return finalConfig;
}
