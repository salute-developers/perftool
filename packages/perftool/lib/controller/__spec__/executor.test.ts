import { jest } from '@jest/globals';

import { Config } from '../../config';
import type { Test } from '../executor';
import Cache from '../../cache';

function createFakeCache(): Cache {
    return {
        getTaskState: jest.fn(() => {}),
        setTaskState: jest.fn(() => {}),
    } as unknown as Cache;
}

describe('controller/Executor', () => {
    afterEach(() => {
        jest.restoreAllMocks();
        jest.resetModules();
    });

    it('should create Executor instance with factory', async () => {
        let launchMock = {} as jest.Mock;
        jest.unstable_mockModule('puppeteer', () => ({
            default: {
                launch: (launchMock = jest.fn(async () => {})),
            },
        }));

        const { default: Executor } = await import('../executor');

        const port = Math.trunc(Math.random() * 2 ** 16);
        const config = {
            puppeteerOptions: {
                waitForInitialPage: true,
                devtools: true,
            },
        } as Config;
        const cache = createFakeCache();

        const executor = await Executor.create(config, cache, port);

        expect(executor).toBeInstanceOf(Executor);
        expect(launchMock).toHaveBeenCalledTimes(1);
        expect(launchMock.mock.calls[0]).toMatchSnapshot();
    });

    it('should open page in browser and return result from client runner', async () => {
        let newPageMock = {} as jest.Mock;
        let gotoMock = {} as jest.Mock;
        let exposeFunctionMock = {} as jest.Mock<any>;
        let closeMock = {} as jest.Mock;
        let bootstrapTestMock = {} as jest.Mock;
        let useInterceptApiMock = {} as jest.Mock;
        let useViewportApiMock = {} as jest.Mock;
        const test: Test = { subjectId: 'fakeId3', taskId: 'fakeTask3', type: 'dry' };
        const result = { ...test, state: {} };
        jest.unstable_mockModule('puppeteer', () => ({
            default: {
                launch: async () => ({
                    newPage: (newPageMock = jest.fn(async () => ({
                        goto: (gotoMock = jest.fn()),
                        exposeFunction: (exposeFunctionMock = jest.fn((_: any, callback: (r: any) => void) => {
                            callback(result);
                        })),
                        close: (closeMock = jest.fn()),
                    }))),
                }),
            },
        }));
        jest.unstable_mockModule('../clientScript', () => ({
            bootstrapTest: (bootstrapTestMock = jest.fn()),
        }));
        jest.unstable_mockModule('../../api/intercept', () => ({
            useInterceptApi: (useInterceptApiMock = jest.fn()),
        }));
        jest.unstable_mockModule('../../api/viewport', () => ({
            useViewportApi: (useViewportApiMock = jest.fn()),
        }));

        const { default: Executor } = await import('../executor');

        const port = Math.trunc(Math.random() * 2 ** 16);
        const config = {
            puppeteerOptions: {},
            runWaitTimeout: 1000,
        } as Config;
        const cache = createFakeCache();

        const executor = await Executor.create(config, cache, port);
        const execResult = await executor.execute(test);

        expect(newPageMock).toHaveBeenCalledTimes(1);

        expect(gotoMock).toHaveBeenCalledTimes(1);
        expect(gotoMock).toHaveBeenCalledWith(`http://localhost:${port}/`);

        expect(exposeFunctionMock).toHaveBeenCalledTimes(2);
        expect(exposeFunctionMock.mock.calls[0][0]).toEqual('_perftool_finish');
        expect(exposeFunctionMock.mock.calls[1][0]).toEqual('_perftool_on_error');

        expect(useInterceptApiMock).toHaveBeenCalledTimes(1);
        expect(useViewportApiMock).toHaveBeenCalledTimes(1);

        expect(bootstrapTestMock).toHaveBeenCalledTimes(1);
        expect(bootstrapTestMock.mock.calls[0][1]).toEqual({ ...test, state: {} });

        expect(closeMock).toHaveBeenCalledTimes(1);

        expect(execResult).toEqual(result);
    });

    it('should return error on page timeout', async () => {
        let closeMock = {} as jest.Mock;
        jest.unstable_mockModule('puppeteer', () => ({
            default: {
                launch: async () => ({
                    newPage: jest.fn(async () => ({
                        goto: jest.fn(),
                        exposeFunction: jest.fn(),
                        addScriptTag: jest.fn(),
                        close: (closeMock = jest.fn()),
                    })),
                }),
            },
        }));
        jest.unstable_mockModule('../clientScript', () => ({
            bootstrapTest: jest.fn(),
        }));
        jest.unstable_mockModule('../../api/intercept', () => ({
            useInterceptApi: jest.fn(),
        }));

        const { default: Executor } = await import('../executor');

        const port = Math.trunc(Math.random() * 2 ** 16);
        const config = {
            puppeteerOptions: {},
            runWaitTimeout: 0,
        } as Config;
        const cache = createFakeCache();

        const executor = await Executor.create(config, cache, port);
        const execResult = await executor.execute({ subjectId: 'fakeId3', taskId: 'fakeTask3' });

        expect(closeMock).toHaveBeenCalledTimes(1);
        expect(execResult).toBeInstanceOf(Error);
    });

    it('should throw if execute called after finalize', async () => {
        let newPageMock = {} as jest.Mock;
        let closeMock = {} as jest.Mock;
        jest.unstable_mockModule('puppeteer', () => ({
            default: {
                launch: async () => ({
                    newPage: (newPageMock = jest.fn()),
                    close: (closeMock = jest.fn()),
                }),
            },
        }));

        const { default: Executor } = await import('../executor');

        const port = Math.trunc(Math.random() * 2 ** 16);
        const config = {
            puppeteerOptions: {},
        } as Config;
        const cache = createFakeCache();

        const executor = await Executor.create(config, cache, port);

        await executor.finalize();
        expect(closeMock).toHaveBeenCalledTimes(1);

        expect(executor.execute({ subjectId: 'fakeId3', taskId: 'fakeTask3', type: 'dry' })).rejects.toThrow();
        expect(newPageMock).toHaveBeenCalledTimes(0);

        await executor.finalize();
        expect(closeMock).toHaveBeenCalledTimes(1);
    });
});
