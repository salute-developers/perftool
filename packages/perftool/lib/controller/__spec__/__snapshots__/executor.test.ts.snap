// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`controller/Executor should create Executor instance with factory 1`] = `
[
  {
    "args": [
      "--js-flags="--maglev=false --max_opt=0"",
      "--no-sandbox",
    ],
    "devtools": true,
    "headless": "new",
    "waitForInitialPage": true,
  },
]
`;
